### PR TITLE
Prototype/state space effector

### PR DESCRIPTION
**Pull request by joaogvcarneiro**
* **Tickets addressed:** #297 
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description

The first commit adds module files, while the second and third commits add test files.

## Verification

Verification and validation test scripts are added to test energy and rigid body motion.

## Documentation

N/A.

## Future work

N/A.

### CODE REVIEW

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp
Comment (patkenneally) [inline src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp]: :small_red_triangle_down: This needs to be updated to read "Copyright (c) 2024 Laboratory for Atmospheric and Space Physics, University ..." You can do this for all files added.

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h
Comment (patkenneally) [inline src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h]: :small_red_triangle_down: Let's mark these all as `override`

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp
Comment (patkenneally) [inline src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp]: :small_red_triangle_down: Copy and paste comment?

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp
Comment (codyallard28) [inline src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp]: Change this to be LASP

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp
Comment (codyallard28) [inline src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp]: Should the default for M and K be diagonal ones?

Comment (codyallard28) [reply src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp]: Looking further down, M does get an inverse so I think it makes sense for M

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp
Comment (codyallard28) [inline src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp:24]: Should every dynamic Eigen matrix from the header file be set to zero or one in the constructor?

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h
Comment (codyallard28) [inline src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h]: Same as above

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.i
Comment (codyallard28) [inline src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.i]: Same as above

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp
Comment (codyallard28) [inline src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp]: I think a comment above each chunk of code in update contributions, compute derivatives, update energy contributions and compute state space energy seems warranted

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp
Comment (patkenneally) [inline src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp]: :large_orange_diamond: Can drop each parameter down on a new line so the line is within 120 characters.

File: src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py
Comment (patkenneally) [inline src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py]: :small_red_triangle_down: Delete this comment block. It duplicates information contained in the git log.

File: src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py
Comment (patkenneally) [inline src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py:32]: :small_red_triangle_down: Merge this function with the actual test function. This separation is a legacy pattern that we are moving away from.

File: src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_validation.py
Comment (patkenneally) [inline src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_validation.py]: :small_red_triangle_down: Delete comment block

File: src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_validation.py
Comment (patkenneally) [inline src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_validation.py:32]: :small_red_triangle_down: Collapse these two functions.

**Review by patkenneally — APPROVED** [review]
Legend:
:small_red_triangle_down: *Issues* to address before merge
:large_orange_diamond: *Requests* that should not block merge, but should at least be discussed
:large_blue_circle: *Recommendations* that can be ignored if desired

Nice! Update all copyright statements to reflect the correct author. Small formatting tweaks.

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h
Comment (codyallard28) [inline src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h]: Does this need to be a SysModel? And if it does, do you need to populate UpdateState and Reset, or can you just leave them as the parent class implementation?

File: src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py
Comment (codyallard28) [inline src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py]: Same

File: src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py
Comment (codyallard28) [inline src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py:107]: Can you make these K's and C's just all slightly different?

**Review by codyallard28 — CHANGES_REQUESTED** [review]
My only show stoppers are initializing your variables to proper values that won't blow up the sim.

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp]: Done

Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp]: Done

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h]: Done

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.i
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.i]: Done

File: src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py]: Done

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp]: Because of the inversion, I'm changing the M matrix to be ``ones`` instead of ``zeros``. This should make the module work without any setup, even if K is ``zeros``. Nice catch!

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp:24]: Discussed above.

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp]: I think the code speaks for itself, but I'll add the comments nonetheless.

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h]: It doesn't need to be a ``SysModel`` to run, but I use the ``logger`` function to grab the state variables for plotting, which is only accessible through the ``SysModel`` class. I don't see an alternative to plotting private variables, and I don't think making them public is a good idea either. However, I'm also not a fan of having a post-processing use case influence how a module is run... @patkenneally can you help a man out here?

File: src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py:107]: Done.

File: src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py]: Done.

File: src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_validation.py
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_validation.py]: Done.

File: src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_validation.py
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_validation.py:32]: Done.

File: src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/_UnitTest/test_stateSpaceStateEffector_verification.py:32]: Done.

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp]: Done.

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.cpp]: Yes, removed it.

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h
Comment (joaogvcarneiro) [reply src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h]: Removed unnecessary function declarations and added ``override`` to the appropriate functions.

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h
Comment (codyallard28) [reply src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h]: I am fine with it staying a SysModel

**Review by codyallard28 — APPROVED** [review]

File: src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h
Comment (patkenneally) [reply src/simulation/dynamics/stateSpaceEffector/stateSpaceStateEffector.h]: @joaogvcarneiro what you have is sufficient for now. Changes coming will remove this multiple coupling of logging, sys_model, etc. We can update this after those changes.

**Pull request merged**
